stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -2,30 +2,315 @@
+
+    @JsonProperty(required = false)
+    public RGBAColor paddingColor;
+
+    @JsonProperty(required = false)
+    public RGBAColor borderColor;
+
+    @JsonProperty(required = false)
+    public RGBAColor marginColor;
+  }
+
+  public static class RGBAColor {
+    @JsonProperty(required = true)
+    public int r;
+
+    @JsonProperty(required = true)
+    public int g;
+
+    @JsonProperty(required = true)
+    public int b;
+
+    @JsonProperty(required = false)
+    public Double a;
+  }
+
+  private void unhookElement(Object element) {
+    synchronized (mNodeMapSync) {
+      int nodeId = mElementToNodeIdMap.get(element);
+      mElementToNodeIdMap.remove(element);
+      mNodeIdToElementMap.remove(nodeId);
+    }
+
+    NodeDescriptor nodeDescriptor = mDOMProvider.getNodeDescriptor(element);
+    nodeDescriptor.unhook(element);
+    int childCount = nodeDescriptor.getChildCount(element);
+    for (int i = 0; i < childCount; ++i) {
+      Object childElement = nodeDescriptor.getChildAt(element, i);
+      NodeDescriptor childNodeDescriptor = mDOMProvider.getNodeDescriptor(childElement);
+      childNodeDescriptor.unhook(childElement);
+    }
+  }
+
+  private Object getElementForNodeId(int nodeId) {
+    synchronized (mNodeMapSync) {
+      return mNodeIdToElementMap.get(nodeId);
+    }
+  }
+
+  private int getOrCreateNodeIdForElement(Object element) {
+    synchronized (mNodeMapSync) {
+      if (mElementToNodeIdMap.containsKey(element)) {
+        return mElementToNodeIdMap.get(element);
+      }
+
+      Integer nodeId = mNextNodeId++;
+      mElementToNodeIdMap.put(element, nodeId);
+      mNodeIdToElementMap.put(nodeId, element);
+
+      NodeDescriptor nodeDescriptor = mDOMProvider.getNodeDescriptor(element);
+      nodeDescriptor.hook(element);
+
+      return nodeId;
+    }
+  }
+
+  private Node createNodeForElement(Object element) {
+    NodeDescriptor nodeDescriptor = mDOMProvider.getNodeDescriptor(element);
+
+    Node node = new Node();
+    node.nodeId = getOrCreateNodeIdForElement(element);
+    node.nodeType = nodeDescriptor.getNodeType(element).getValue();
+    node.nodeName = nodeDescriptor.getNodeName(element);
+    node.localName = nodeDescriptor.getLocalName(element);
+    node.nodeValue = nodeDescriptor.getNodeValue(element);
+
+    node.children = getChildNodesForElement(element);
+    node.childNodeCount = node.children.size();
+
+    int attributeCount = nodeDescriptor.getAttributeCount(element);
+    if (attributeCount > 0) {
+      node.attributes = new ArrayList<>(attributeCount * 2);
+      NodeAttribute attribute = new NodeAttribute();
+
+      for (int i = 0; i < attributeCount; ++i) {
+        nodeDescriptor.getAttributeAt(element, i, attribute);
+        if (attribute.name == null || attribute.value == null) {
+          throw new IllegalArgumentException();
+        }
+
+        node.attributes.add(attribute.name);
+        node.attributes.add(attribute.value);
+
+        attribute.name = null;
```

Comment:
Defensive programming is anything but here.  It would lead to subtle bugs going undetected because new fields might not be copied but the caller would expect that they were.  Document that all fields must be set so it's clear that a bug exists if this happens in practice.